cluster-autoscaler/utils/taints/taints.go

```diff
@@ -18,17 +24,42 @@ func hasNidhoggTaint(node *apiv1.Node) bool {
 	return false
 }
 
-func FilterOutNodesWithTaints(allNodes, readyNodes []*apiv1.Node) ([]*apiv1.Node, []*apiv1.Node) {
+func FilterOutNodesWithTaints(kubeClient kubernetes.Interface, allNodes, readyNodes []*apiv1.Node) ([]*apiv1.Node, []*apiv1.Node) {
 	var newAllNodes []*apiv1.Node
 	var newReadyNodes []*apiv1.Node
 	nodesWithTaints := make(map[string]*apiv1.Node)
 
 	for _, node := range readyNodes {
+		_, ok := node.Annotations[nidhoggTaintPrefix+"/ready"]
+		if ok {
+			newReadyNodes = append(newReadyNodes, node)
+		}
 		if hasNidhoggTaint(node) {
 			nodesWithTaints[node.Name] = utils.GetUnreadyNodeCopy(node)
-		} else {
-			newReadyNodes = append(newReadyNodes, node)
+			continue
+		}
+		oriB, err := json.Marshal(node)
+		if err != nil {
+			klog.Errorf("failed to marshal patch for node %s to annotate: %v", node.Name, err)
+			continue
+		}
+		node.Annotations[nidhoggTaintPrefix+"/ready"] = time.Now().Format("2006-01-02T15:04:05Z")
+		modB, err := json.Marshal(node)
+		if err != nil {
+			klog.Errorf("failed to marshal patch for node %s to annotate: %v", node.Name, err)
+			continue
+		}
+		patch, err := strategicpatch.CreateTwoWayMergePatch(oriB, modB, apiv1.Node{})
+		if err != nil {
+			klog.Errorf("failed to create patch for node %s to annotate: %v", node.Name, err)
+			continue
+		}
+		_, err = kubeClient.CoreV1().Nodes().Patch(node.Name, types.StrategicMergePatchType, patch)
```

Comment:
I have mixed feelings about this. 
This function `FilterOutNodesWithTaints` is suddenly doing way more than what it suggests: 
1) It isn't just filtering on taints any more, we are filtering based on annotations as well
2) It isn't just filtering, it is patching nodes too.

While I get the reason for doing this, I'm unsure whether we should add the annotation `nidhoggTaintPrefix+"/ready"` here. How about we fork nidhogg instead and add this annotation there?
I know it sounds ridiculous to suggest to fork one more repo, but I feel like that's a better compromise than the coupling between components here. And, it will make it harder to debug/understand as well.

Also, can we maybe think of solution without this `ready` annotation? Like maybe have cluster autoscaler emit a metric for number of unready nodes due to nidhogg taints and alert us when it reaches close to a limit that will impact the cluster autoscaler?

I don't feel like my suggestions are significantly better than what you are doing here either